test/old_tests/UnitTests/References.cpp

```diff
@@ -27,4 +28,44 @@ TEST_CASE("References")
 
     // Tests short circuit when one is empty.
     REQUIRE(a != e);
+
+    // Different integer types, but same value
+    REQUIRE(c == f);
+
+    // Different integer types, but one is null.
+    REQUIRE(e != f);
+
+    // Tests comparison to value.
+    REQUIRE(a == 123);
+    REQUIRE(123 == a);
+
+    REQUIRE(a != 124);
+    REQUIRE(124 != a);
+
+    // Tests short circuit when reference is empty.
+    REQUIRE_FALSE(e == 123);
+    REQUIRE_FALSE(123 == e);
+
+    REQUIRE(e != 123);
+    REQUIRE(123 != e);
```

Comment:
Let's also add tests for floats, doubles, and bools. Testing against bool in particular will also be good for weeding out any surprises, most likely around types that are convertible to bool. It makes me wonder if we are going to inadvertently allow things like this to compile:
```
IReference<bool> == L"foo";
IReference<bool> == 42;
IReference<bool> == IInspectable{};
```
etc

Reply:
It also means that if `foo` is an `IReference<bool>`, then

```cpp
if (foo != false) ...
if (foo) ...
```
are not the same thing, which could be the source of subtle refactoring bugs where somebody takes the rule "Don't compare a boolean against true or false, just test the boolean" and inadvertently applies it to the first version to "simplify" it to the second version, and ends up breaking it.

Reply:
Yes, we need to be very careful not to introduce a breaking change here.

Reply:
Raymond raises a valid point, and you can see this in how C# handles a `bool?`: you can't test those just by doing `if (foo)`, it results in a compilation error, you have to write `if (foo == true)`, making it effectively a tri-state bool. Doing this in C++ would be possible but it would obviously be source breaking.

Reply:
FWIW I think `if (foo)` is already misleading because it only tests nullability, not if it's actually true.

Reply:
Other things that would be not be equivalent after this change:

* `if (!v)` and `if (v == false)`
* `if (v)` and `if (v != 0)`

The case of `v != 0` is problematic because `0` is implicitly convertible to `nullptr`, so a test like `if (v != 0)` does different things depending on whether `v` is `IReference<int>` or `IReference<not-int>`.

I admire the intent behind these convenience operators, but I think they open too big of a can of worms.

Reply:
> I admire the intent behind these convenience operators, but I think they open too big of a can of worms.

Agreed, C++ lets you express some amazing things but it can often get you into trouble. In this case, we have a *lot* of code that has to keep running on C++/WinRT and I'm just not sure the convenience here is worth the risk.

Reply:
Right, and it gets worse with the `NULL` macro being actually defined to 0 so if somebody uses that to check for null with an `IReference<int>`, this would actually be breaking. If I really want this feature, I can implement those operators manually in my code.

I still think there might be value in merging the `std::wstring_view` overloads and ability to compare different types however, provided I address @DefaultRyan's comment.

Reply:
I think I'm with Raymond and Kenny here. Adding overloads for `operator==` to pre-existing types is a change that is fundamentally invasive to existing code - code that was testing/comparing `IReference<T>` may now acquire very different semantics, because `operator==` is fundamental to the language and unavoidable.

Adding a new helper function (and to a lesser degree, modifying an existing helper) is far less invasive, because it allows for more "opt in" on the developers' part.

But in either case, the benefit of this change seems relatively meager. All it enables is replacing this
```
if (ref && ref.Value() == value)
```
with this
```
if (ref == value)
```

The old pattern isn't onerous, hostile, or overly verbose, so the convenience almost borders on syntactic sugar. Like Ramond, I admire the intent, but the edge cases are a bit too subtle and potentially breaking in my opinion.

Reply:
This might be more suited to a helper library like [`wil\include\cppwinrt.h`](https://github.com/microsoft/wil/blob/master/include/wil/cppwinrt.h).

```cpp
namespace wil
{
   template<typename T, typename T>
    bool nullable_value_equal(IReference<T> const& left, T const& right)
    {
        return left && left.Value() == right;
    }

   // etc.
}
```

Then people could opt into it explicitly:

```cpp
  if (wil::nullable_value_equal(foo, true)) { ... }
```